File: examples/ts/intent-current-block-height/IntentCurrentBlockHeight.runar.ts
import { StatefulSmartContract, assert, currentBlockHeight } from 'runar-lang';

/**
 * IntentCurrentBlockHeight exercises the `currentBlockHeight`
 * shorthand, which is pure source-level sugar for
 * `extractLocktime(this.txPreimage)`. The desugar happens at ANF
 * lowering time -- no new ANF kind or stack codegen is needed.
 */
class IntentCurrentBlockHeight extends StatefulSmartContract {
  readonly deadline: bigint;
  count: bigint;

  constructor(deadline: bigint, count: bigint) {
    super(deadline, count);
    this.deadline = deadline;
    this.count = count;
  }

  public spend() {
    const h = currentBlockHeight();
    assert(h <= this.deadline);
    this.count = this.count + 1n;
  }
}

File: examples/python/intent-current-block-height/IntentCurrentBlockHeight.runar.py
from runar import (
    StatefulSmartContract, Bigint, Readonly,
    public, assert_, current_block_height,
)


class IntentCurrentBlockHeight(StatefulSmartContract):
    """Exercises the current_block_height shorthand, which is pure
    source-level sugar for extract_locktime(self.tx_preimage)."""

    deadline: Readonly[Bigint]
    count: Bigint

    def __init__(self, deadline: Bigint, count: Bigint):
        super().__init__(deadline, count)
        self.deadline = deadline
        self.count = count

    @public
    def spend(self):
        h = current_block_height()
        assert_(h <= self.deadline)
        self.count = self.count + 1
